### DISCUSSION TITLE
How To Make A Feature Request?

### DISCUSSION THREAD

**Discussion by TheMakerTales** [hlorus/CAD_Sketcher#106] [Announcements]
We are always happy to hear your feature requests.

Please follow the instructions below to make your request is as informative and compelling as possible. This will give other users a reason to upvote your request and attract developers who may want to attempt your feature request or add it to upcoming releases. 

 
 
### 1. Search & Ask
Take a good look through the [roadmap](https://github.com/users/hlorus/projects/1) and the other [requests](https://github.com/hlorus/CAD_Sketcher/discussions/categories/feature-requests) to make sure you are not making a duplicate request. _Duplicate requests will be deleted._ Remember that upvoting and adding more information to ongoing requests is much more valuable to our community. If you have any doubts feel free to bring it up in the [Discord](https://discord.gg/EyRjypztmf) first.

 
 
### 2. Describe & Illustrate your Feature Request
The more information you can provide the more likely your feature request will get picked up by the community and developers. Even if the title says it all do your best to fill in the information below:

 
 
_Copy and paste the text below into your request:_

**Motivation / Existing Problem**

**Description**

**Alternative Solutions**

**Mockup**
Ideally, add a quick drawing to show how this would look.

 
 
### 3. Let us know all about it
Once you have posted your feature request let the [Discord](https://discord.gg/EyRjypztmf) community know about it. Usually, the best place is in the general discussion channel.

**Comment by SmallTownSound** [comment]
Feature Request:
Insert entity at solver point
Motivation / Existing Problem
A clear and concise description of what the problem is. Ex. I'm always frustrated when [...]
It is a pain to delete points and lines when you got a change in the part, such as a rounded corner when it was just 2 lines joining before. 
Description
A clear and concise description of what you want to happen.
When right clicking on a Solvespace point, a choice in the list would be to insert an entity at this point.  Obviously, if a curve center point is selected, the 'add' option would not be given.
Alternative Solutions
A clear and concise description of any alternative solutions or features you've considered.

Mockup
Ideally, add a quick drawing to show how this would look.
Just add to the right click options.

**Comment by abergmeier** [comment]
**Motivation:**
Increase parallelism when solving a Sketch.

**Description:**
It should be possible to use multiple `Solver` instances for a Sketch, when not all Entities are connected in the Constraint Graph.

**Mockup:**
Steps necessary:
1. Partitionize Constraint Graph
2. Create `Solver` instances for every partition
3. Run solve in parallel

Optional steps:
1.1. Cache partitions

**Comment by mac4kent** [comment]
Feature this CAD Sketch on Blender 3.2.2 for Mac. Do not use Windows.
I see you made Keymap code on CAD Sketch and I am show you screenshot on Mac.
https://imgur.com/9FiQte1

Your code is https://imgur.com/n0Y6xSh

https://support.apple.com/en-us/HT201236

**Comment by Norren** [comment]
This one's just a little request, but could you please start posting the current version number and its release date next to the Zip download link on the Gumroad page?

**Comment by mac4kent** [comment]
Oh well I do use Mac. I do not use Windows sorry.
They ignore me about Mac.

**Comment by Cook-droid** [comment]
希望能有主动选择显示或者隐藏草图的功能，这样能知道在哪个平面进行工作

**Comment by mac4kent** [comment]
I don’t understand you

**Comment by ode2spot** [comment]
**Motivation / Existing Problem**
need to add unnecessary construction lines to set horizontal/vertical (orthogonal) distance constraints.  This leads to extra items in sketches and is hard to maintain

**Description**
It would be really helpful to have two additional constraint types: vertical distance, and horizontal distance.

**Alternative Solutions**
I am currently getting around this by adding extra construction lines.  This adds noise to my workflow and is hard to update.

**Comment by tobiasBora** [comment]
**Motivation / Existing Problem**

I’d love to use CAD sketcher to create some animations by changing the shape of a figure, by allowing the user to add **keyframes** on the drawing (e.g. on length). 

**Description**
Ideally most inputs should be keyframable. If the time to render is too long, caching/baking may help.

**Alternative Solutions**
Convert back to mesh, but lose advantages of CAD. Maybe shape keys can be a solution for simpler evolution (proposed on discord), but I don’t know if I need to convert it to a mesh first or if I can still use CAD features. In any case, shape keys are not as precise for more complex animations.

**Comment by tobiasBora** [comment]
**Motivation / Existing Problem**

Python API and/or geometry node compatibility: there are some shapes I might want to generate with many variations (e.g. number of holes, size, sequence of different shapes converging asymptotically to another shape…) .So a python api and/or a way to integrate them with geometry nodes would be great to have. 

**Description**
The python api could be done so that I can write something like:
```
n = 10
pointsTop = cad.createPoints(range(n))
pointsBottom = cad.createPoints(range(n))
leftLine = cad.addLine(pointsTop[0],pointsBottom[0])
rightLine = cad.addLine(pointsTop[-1],pointsBottom[-1])
arcsTop = [cad.addArc(center = cad.middle(pointsTop[i],pointsTop[i+1]), left=pointsTop[i], right=pointsTop[i+1]) for i in range(n-1)]
for arc in arcsTop[1:]:
    cad.equalRadius(arc, arcTop[0])
…
```
**Alternative Solutions**
Do it without CAD, possible for simple shape but more involved shapes involving tangents might be more challenging to do this way.

**Comment by kayboy-bebop** [comment]
Feature request: Toggle bezier curve direction

When working with Bezier curves it would be very useful to have a way to quickly reverse its direction. At the moment you either have to edit the bezier manually (which will be cleared if sketch is changed) or,  through trial and error, redraw the sketch to get desired direction.

This is relevant when using a bezier as the bevel object for a bezier sketch.
![Screenshot from 2023-10-18 13-31-47](https://github.com/hlorus/CAD_Sketcher/assets/115389530/b1e3c976-224a-4de1-88f7-01ebf9460180)

**Comment by raindropsfromsky** [comment]
I think it is better to let the Issue tracker collect feature requests. GitHub allows you to change the structure of each category (bug vs feature request).  Compiling all feature requests in a single thread would create a mess!

**Comment by UltraBlackLinux** [comment]
Add a "duplicate sketch" button
Motivation: Copy-Pasting is unreliable and moves sketches by default.
Description: A "duplicate sketch" button
Alternative solutions: none
Mockup: RMB -> Duplicate sketch

**Comment by RoboRobob** [comment]
Request for the addition of Ellipses and Elliptical curves.

Motivation / Existing Problem
I have not seen any indication of the ability to add ellipses or elliptical curves to Cad sketcher. In my current project I was trying to find an indication of the ability to use one in a sketch and I fond nothing in the documentation or the github page discussing them, so I assume they are currently not on the roadmap.

Description
I would imagine a button hidden behind the circle and curve buttons that would change the type of curve/circle to an ellipse, but I'm not a UI/UX designer so there may be better options than that for forward compatibility. 

A clear and concise description of what you want to happen.
The idea is - at a top down level, that it would work similarly to how the circle and curve tools work, just expanded as a new tool type.

Comments:
This may be labor intensive, or it may not. I've never worked on a feature like this so i don't know. Thank you for your time.

Mockup
![ellipse_standard_equation_1-2762406425](https://github.com/user-attachments/assets/0ca14a4d-9cee-4793-9708-2d6627754f7a)

**Comment by Isaina** [comment]
Hi can you add like this editing and make library of reusable components for just drag and drop 
<img width="555" height="499" alt="IMG_2986" src="https://github.com/user-attachments/assets/b0b5daee-ff4f-49d4-b45e-070d5270f32c" />

**Comment by skrzypens** [comment]
In 0.3 in sketch edit mode moving the point result in resizing of the geometry. I may be missing something but feels to me it would be nicer if point grab moves the geometry and lines, arches resize it. 
Would be nice if snap markers work when moving it (geometry) not only when new ones are created.

**Reply by hlorus** [reply]
Hey, snapping now works also while dragging existing points.
